### ISSUE TITLE
to_string and from_string have different behavior for TableReference

### ISSUE THREAD

**Issue by waynexia** [apache/datafusion#6853]
### Describe the bug

`to_string` simply concat components with `.`:
https://github.com/apache/arrow-datafusion/blob/02a470f6061cce8ee8e57f7af8a6a0e0ddc1571b/datafusion/common/src/table_reference.rs#L107-L121

but `from_string` will transform them to lowercase:
https://github.com/apache/arrow-datafusion/blob/02a470f6061cce8ee8e57f7af8a6a0e0ddc1571b/datafusion/common/src/utils.rs#L248-L257

`TableReference` is the most common table identifier for tables. But I don't think converting cases or considering quotes is necessary when the object name becomes `TableReference`. At this stage (i.e., after the initial plan is constructed from AST), those object names should be fixed.

This implicit conversion makes it difficult when handling complex table identifiers. Thus I propose to make `TableReference` as simple as possible and gather all those `to_lower_cases` and "quotes" before generating the very first `TableReference`.

### To Reproduce

```rust
    #[test]
    fn to_lower_case() {
        let table_ref = TableReference::Full {
            catalog: Cow::Owned("catalog".to_string()),
            schema: Cow::Owned("schema".to_string()),
            table: Cow::Owned("TABLE".to_string()),
        };
        let concatted = table_ref.to_string();
        println!("concatted: {}", concatted);
        let parsed = TableReference::parse_str(&concatted);
        println!("parsed: {:?}", parsed.to_vec());
    }
```

The output is
```
concatted: catalog.schema.TABLE
parsed: ["catalog", "schema", "table"]
```

### Expected behavior

```
concatted: catalog.schema.TABLE
parsed: ["catalog", "schema", "TABLE"]
```

### Additional context

(Not sure if this is "bug"...)

**Comment by Jefffrey** [comment]
Another option is to replace the code in the display impl with `to_quoted_string()`

https://github.com/apache/arrow-datafusion/blob/d6985b0fc1e4362b80beea059857c6bde65b4d04/datafusion/common/src/table_reference.rs#L269-L297

Which should minimally quote the output identifier string, preserving roundtrip ability between `to_string()` and `from_string()`

>  Thus I propose to make `TableReference` as simple as possible and gather all those `to_lower_cases` and "quotes" before generating the very first `TableReference`.

Could you expand on this point a bit? My only concern is that if an identifier part contains a period `.` then the quotes would still be necessary in the output of `to_string()`, e.g. `catalog."data.base".table`

**Comment by Jefffrey** [comment]
Yeah, I guess a good first step is refactoring those usages in tests to make the construction of TableReferences more explicit and narrow down actual usages of the parse from string.
